components/canvas/canvas_data.rs

```diff
@@ -470,12 +471,44 @@ impl<'a> CanvasData<'a> {
             return; // Paint nothing if gradient size is zero.
         }
 
-        let draw_rect = Rect::new(
-            rect.origin,
-            self.backend
-                .size_from_pattern(&rect, &self.state.fill_style)
-                .unwrap_or(rect.size),
-        );
+        let draw_rect = match &self.state.fill_style {
+            Pattern::Raqote(crate::raqote_backend::Pattern::Surface(pattern)) => {
+                let pattern_rect = Rect::new(Point2D::origin(), pattern.size());
+                let mut draw_rect = rect.intersection(&pattern_rect).unwrap_or(Rect::zero());
+
+                let limitx = |draw_rect: &mut Rect<f32>, pattern_rect: &Rect<f32>| {
+                    if draw_rect.size.width > pattern_rect.size.width {
+                        draw_rect.size.width = pattern_rect.size.width;
+                    }
+                };
+                let limity = |draw_rect: &mut Rect<f32>, pattern_rect: &Rect<f32>| {
+                    if draw_rect.size.height > pattern_rect.size.height {
+                        draw_rect.size.height = pattern_rect.size.height;
+                    }
+                };
+
+                match pattern.repeat {
+                    Repetition::NoRepeat => {
+                        limitx(&mut draw_rect, &pattern_rect);
+                        limity(&mut draw_rect, &pattern_rect);
+                    },
+                    Repetition::RepeatX => {
+                        draw_rect.size.width = rect.size.width;
+                        limity(&mut draw_rect, &pattern_rect);
+                    },
+                    Repetition::RepeatY => {
+                        draw_rect.size.height = rect.size.height;
+                        limitx(&mut draw_rect, &pattern_rect);
+                    },
+                    Repetition::Repeat => {
+                        draw_rect = *rect;
+                    },
+                }
+
+                draw_rect
+            },
+            _ => *rect,
```

Comment:
Do we actually need this catch-all branch any more?

Reply:
Well, there are more variants in `raqote_backend::Pattern`, so rustc friendly reminds me that those aren't handled. I rewrote the whole match to be more explicit.